Mathlib/Analysis/CStarAlgebra/Classes.lean

```diff
@@ -0,0 +1,126 @@
+  mul_comm := mul_comm
+
+instance (priority := 100) NonUnitalCommCStarAlgebra.toNonUnitalNormedCommRing
+    (A : Type*) [NonUnitalCommCStarAlgebra A] : NonUnitalNormedCommRing A where
+  mul_comm := mul_comm
+
+/-- This is not registered as an instance to avoid Lean searching for `IsClosed (s : Set A)`
+instances frequently. -/
+@[reducible]
+noncomputable def StarSubalgebra.cstarAlgebra {S A : Type*} [CStarAlgebra A]
+    [SetLike S A] [SubringClass S A] [SMulMemClass S ℂ A] [StarMemClass S A]
+    (s : S) (h_closed : IsClosed (s : Set A)) : CStarAlgebra s where
+  toCompleteSpace := h_closed.completeSpace_coe
+  norm_mul_self_le x := CStarRing.norm_star_mul_self (x := (x : A)) |>.symm.le
+
+/-- This is not registered as an instance to avoid Lean searching for `IsClosed (s : Set A)`
+instances frequently. -/
+@[reducible]
+noncomputable def StarSubalgebra.commCStarAlgebra {S A : Type*} [CommCStarAlgebra A]
+    [SetLike S A] [SubringClass S A] [SMulMemClass S ℂ A] [StarMemClass S A]
+    (s : S) (h_closed : IsClosed (s : Set A)) : CommCStarAlgebra s where
+  toCompleteSpace := h_closed.completeSpace_coe
+  norm_mul_self_le x := CStarRing.norm_star_mul_self (x := (x : A)) |>.symm.le
+  mul_comm _ _ := Subtype.ext <| mul_comm _ _
+
+/-- This is not registered as an instance to avoid Lean searching for `IsClosed (s : Set A)`
+instances frequently. -/
+@[reducible]
+noncomputable def NonUnitalStarSubalgebra.nonUnitalCStarAlgebra {S A : Type*}
+    [NonUnitalCStarAlgebra A] [SetLike S A] [NonUnitalSubringClass S A] [SMulMemClass S ℂ A]
+    [StarMemClass S A] (s : S) (h_closed : IsClosed (s : Set A)) : NonUnitalCStarAlgebra s where
+  toCompleteSpace := h_closed.completeSpace_coe
+  norm_mul_self_le x := CStarRing.norm_star_mul_self (x := (x : A)) |>.symm.le
+
+/-- This is not registered as an instance to avoid Lean searching for `IsClosed (s : Set A)`
+instances frequently. -/
+@[reducible]
+noncomputable def NonUnitalStarSubalgebra.nonUnitalCommCStarAlgebra {S A : Type*}
+    [NonUnitalCommCStarAlgebra A] [SetLike S A] [NonUnitalSubringClass S A] [SMulMemClass S ℂ A]
+    [StarMemClass S A] (s : S) (h_closed : IsClosed (s : Set A)) : NonUnitalCommCStarAlgebra s where
+  toCompleteSpace := h_closed.completeSpace_coe
+  norm_mul_self_le x := CStarRing.norm_star_mul_self (x := (x : A)) |>.symm.le
+  mul_comm _ _ := Subtype.ext <| mul_comm _ _
+
+noncomputable instance : CommCStarAlgebra ℂ where
+  mul_comm := mul_comm
+
+section Pi
+
+variable {ι : Type*} {A : ι → Type*} [Fintype ι]
+
+instance [(i : ι) → NonUnitalCStarAlgebra (A i)] : NonUnitalCStarAlgebra (Π i, A i) where
+
+instance [(i : ι) → NonUnitalCommCStarAlgebra (A i)] : NonUnitalCommCStarAlgebra (Π i, A i) where
+  mul_comm := mul_comm
+
+noncomputable instance [(i : ι) → CStarAlgebra (A i)] : CStarAlgebra (Π i, A i) where
+
+noncomputable instance [(i : ι) → CommCStarAlgebra (A i)] : CommCStarAlgebra (Π i, A i) where
+  mul_comm := mul_comm
+
+end Pi
```

Comment:
`ULift` and `PUnit` would be good instances too, assuming the machinery is present elsewhere.